core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -245,9 +245,17 @@ class RequestChannel(val numProcessors: Int, val queueSize: Int) extends KafkaMe
   /** Send a response back to the socket server to be sent over the network */
   def sendResponse(response: RequestChannel.Response) {
     if (isTraceEnabled) {
-      val requestHeader = response.request.header
-      trace(s"Sending ${requestHeader.apiKey} response to client ${requestHeader.clientId} of " +
-        s"${response.responseSend.size} bytes.")
+      response.responseAction match {
+        case SendAction =>
+          val requestHeader = response.request.header
+          trace(s"Sending ${requestHeader.apiKey} response to client ${requestHeader.clientId} of ${response.responseSend.get.size} bytes.")
+        case NoOpAction =>
+          val requestHeader = response.request.header
+          trace(s"No need to send ${requestHeader.apiKey} response to client ${requestHeader.clientId}.")
+        case CloseConnectionAction =>
+          trace("The connection is being closed since request handler encountered an error.")
+        case _ => // it should not happen
```

Comment:
not clear why you shadow impossible case? I would remove it or throw something like IllegalArgurmentException

Reply:
Yes, we should remove it and make `ResponseAction` sealed.